drivers/virtio/virtio-mmio.c

```diff
@@ -24,212 +24,852 @@
 {
-  uint32_t id = idx % virtq->len;
+  FAR struct virtio_mmio_device_s *vmdev =
+    (FAR struct virtio_mmio_device_s *)vdev;
+  metal_io_write32(&vmdev->cfg_io, VIRTIO_MMIO_STATUS, status);
+}
 
-  vrtinfo("virtq=%p, idx=%" PRId32 "\n", virtq, idx);
+/****************************************************************************
+ * Name: virtio_mmio_get_status
+ ****************************************************************************/
+
+static uint8_t virtio_mmio_get_status(FAR struct virtio_device *vdev)
+{
+  FAR struct virtio_mmio_device_s *vmdev =
+    (FAR struct virtio_mmio_device_s *)vdev;
+  return metal_io_read32(&vmdev->cfg_io, VIRTIO_MMIO_STATUS);
+}
+
+/****************************************************************************
+ * Name: virtio_mmio_write_config
+ ****************************************************************************/
+
+static void virtio_mmio_write_config(FAR struct virtio_device *vdev,
+                                     uint32_t offset, void *src, int length)
+{
+  FAR struct virtio_mmio_device_s *vmdev =
+    (FAR struct virtio_mmio_device_s *)vdev;
+  uint32_t write_offset = VIRTIO_MMIO_CONFIG + offset;
+  uint32_t u32data;
+  uint16_t u16data;
+  uint8_t u8data;
+
+  if (vdev->id.version == VIRTIO_MMIO_VERSION_1)
+    {
+      FAR char *s = src;
+      int i;
+      for (i = 0; i < length; i++)
+        {
+          metal_io_write8(&vmdev->cfg_io, write_offset + i, s[i]);
+        }
+
+      return;
+    }
+
+  switch (length)
+    {
+      case 1:
+        memcpy(&u8data, src, sizeof(u8data));
+        metal_io_write8(&vmdev->cfg_io, write_offset, u8data);
+        break;
+      case 2:
+        memcpy(&u16data, src, sizeof(u16data));
+        metal_io_write16(&vmdev->cfg_io, write_offset, u16data);
+        break;
+      case 4:
+        memcpy(&u32data, src, sizeof(u32data));
+        metal_io_write32(&vmdev->cfg_io, write_offset, u32data);
+        break;
+      case 8:
+        memcpy(&u32data, src, sizeof(u32data));
+        metal_io_write32(&vmdev->cfg_io, write_offset, u32data);
+        memcpy(&u32data, src + sizeof(u32data), sizeof(u32data));
+        metal_io_write32(&vmdev->cfg_io, write_offset + sizeof(u32data),
+                         u32data);
+        break;
+      default:
+        DEBUGASSERT(0);
+    }
+}
+
+/****************************************************************************
+ * Name: virtio_mmio_read_config
+ ****************************************************************************/
+
+static void virtio_mmio_read_config(FAR struct virtio_device *vdev,
+                                    uint32_t offset, FAR void *dst,
+                                    int length)
+{
+  FAR struct virtio_mmio_device_s *vmdev =
+    (FAR struct virtio_mmio_device_s *)vdev;
+  uint32_t read_offset = VIRTIO_MMIO_CONFIG + offset;
+  uint32_t u32data;
+  uint16_t u16data;
+  uint8_t u8data;
+
+  if (vdev->id.version == VIRTIO_MMIO_VERSION_1)
```

Comment:
why version 1 read byte by byte?

Reply:
Because it could work for version 1.

Reply:
ADD: it work for version 2 too, but this behavior not follow the standard, I did this follow the Linux does, and from the linux commit:
https://github.com/torvalds/linux/commit/704a0b5f234db26de5203740999e39523cfa4e3a
I guess some modern virtio deivces do not support the byte-to-byte config read.

Reply:
should we use the byte version for size > 8, like this:
```
  FAR struct virtio_mmio_device_s *vmdev =
    (FAR struct virtio_mmio_device_s *)vdev;

  offset += VIRTIO_MMIO_CONFIG;

  if (vdev->id.version == VIRTIO_MMIO_VERSION_1)
    {
       goto version_1;
    }

  switch (length)
    {
      case 1:
        {
          uint8_t data;
          memcpy(&data, src, sizeof(data));
          metal_io_write8(&vmdev->cfg_io, offset, data);
          break;
       }
      case 2:
        {
          uint16_t data;
          memcpy(&data, src, sizeof(data));
          metal_io_write16(&vmdev->cfg_io, offset, data);
          break;
        }
      case 4:
        {
          uint32_t data;
          memcpy(&data, src, sizeof(data));
          metal_io_write32(&vmdev->cfg_io, offset, data);
          break;
       }
      case 8:
        {
          uint32_t data[2];
          memcpy(data, src, sizeof(data));
          metal_io_write32(&vmdev->cfg_io, offset, data[0]);
          metal_io_write32(&vmdev->cfg_io, offset + 4, data[1]);
          break;
        }
      default:
      version1:
        {
          FAR const uinit8_t *data = src;
          int i;

          for (i = 0; i < length; i++)
            {
              metal_io_write8(&vmdev->cfg_io, offset + i, data[i]);
            }

          break;
        }
   }
```

Reply:
Done, I change to
```c
  if (vdev->id.version == VIRTIO_MMIO_VERSION_1 || length > 8)
    {
      FAR char *s = src;
      int i;
      for (i = 0; i < length; i++)
        {
          metal_io_write8(&vmdev->cfg_io, write_offset + i, s[i]);
        }

      return;
    }
```